src/dune_engine/action_builder.ml

```diff
@@ -79,22 +79,22 @@ let env_var s = deps (Dep.Set.singleton (Dep.env s))
 
 let alias a = dep (Dep.alias a)
 
-let contents_impl p =
-  of_thunk
-    { f =
-        (fun _mode ->
-          let open Memo.O in
-          let+ x = Build_system.read_file p ~f:Io.read_file in
-          (x, Dep.Map.empty))
-    }
-
 let contents =
-  let memo =
-    create_memo "file-contents"
-      ~input:(module Path)
-      ~cutoff:String.equal contents_impl
+  let read_file =
+    Memo.exec
+      (Memo.create "Action_builder.contents"
+         ~input:(module Path)
+         ~cutoff:String.equal
+         (fun p -> Build_system.read_file p ~f:Io.read_file))
```

Comment:
Btw, does it make sense to run `Io.read_file` in a background thread?

Reply:
Yeah, it wouldn't be such a bad idea to experiment with it. Although my experience with actions tells me we need to reduce the overhead here first.